### PR TITLE
Fix optimisation of switches to lookup tables

### PR DESCRIPTION
**Pull request by stedolan**
This pull request is remarkably similar to #863, but:
  - removes misguided buggy bit-twiddling
  - adds tests

### CODE REVIEW

**Comment by gasche** [comment]
Would you consider splitting the commit in two, with first the `constant_closures -> cmm_constants` move (which should be semantics-preserving and easier to review) and the tests, and second the optimization? (We could even be very fine, oh so very fine-grained, and have a first commit with only the `ConstClosure` constructor, and `ConstBlock` coming as a second commit.)

(It's not only about the review, which is doable from just the current patch, but also about isolating the optimization, the part that we suspect we may want to temporally revert for debugging or impact study in the future, from the generalization on which even more cool things may be built from in the future.)

Nitpick: CamlCase is not idiomatic for constructor names in this part of the codebase, we rather use separate_words.

**Comment by stedolan** [comment]
Now split in two.

**Comment by stedolan** [comment]
Hmmm, now that I read your middle paragraph I see you had something slightly different in mind (I thought the split was just to make code review easier).

It's now in three commits: refactor; delete the broken optimisation and add tests; add the fixed optimisation

**Comment by gasche** [comment]
As far as I am concerned, this is very nice, thanks. We need a review from someone that understand the backend better than I do, but then I would support merging. Could you edit the relevant Changes entry to also indicate this PR, and 1068 as well?

File: asmcomp/cmmgen.ml
Comment (mshinwell) [inline asmcomp/cmmgen.ml:1361]: This was present in the previous version of this patch too, but anyway: what is the reason for these two tests on the tag bit?  Even if that bit was clear, the values are still constant, so I'm not sure why they can't go in the array.

File: testsuite/tests/basic/switch_opts.ml
Comment (mshinwell) [inline testsuite/tests/basic/switch_opts.ml]: Was the reason for avoiding the constant sharing here just one of keeping the test case self contained?  I think `Sys.opaque_identity` should probably be used.

File: asmcomp/cmmgen.ml
Comment (mshinwell) [inline asmcomp/cmmgen.ml]: Why is this `tag_int` needed?

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: asmcomp/cmmgen.ml
Comment (stedolan) [reply asmcomp/cmmgen.ml:1361]: If the low bit is clear, then the array is not a valid OCaml value. I suppose it doesn't really need to be: I think we can be sure that the GC never touches it, since the address of the array never escapes.

Would it be a good idea to allow the low bit to be zero and delete the block header? Then it would be a plain lookup table, not pretending to be an OCaml heap value.

File: testsuite/tests/basic/switch_opts.ml
Comment (stedolan) [reply testsuite/tests/basic/switch_opts.ml]: I don't think that Sys.opaque_identity would prevent constant sharing here.

File: asmcomp/cmmgen.ml
Comment (stedolan) [reply asmcomp/cmmgen.ml]: The array indexing logic (`Cmmgen.array_indexing`) expects the index to be a tagged integer, but is good at optimising away the tagging. In practice, this generates good code.

File: asmcomp/cmmgen.ml
Comment (mshinwell) [reply asmcomp/cmmgen.ml:1361]: Allowing the bit to be zero and removing the header sounds fine to me.  Or maybe change the tag to `Abstract_tag` instead of removing the header (might be easier)?

File: testsuite/tests/basic/switch_opts.ml
Comment (mshinwell) [reply testsuite/tests/basic/switch_opts.ml]: Hmm, maybe not---although it could be used to stop Flambda evaluating the string concatenation (which it might do in the future).  I'm not quite sure what to do here.  How important is it that the constants aren't shared?  That isn't clear to me.

**Comment by gasche** [comment]
Where are we on this? The patch as it is looks fine to me (I'm not against playing tricks with outside-the-heap values, but I'm not particularly excited either), but in any case I would like the miscompilation currently in trunk to be fixed as soon as possible, which probably means reverting the previous PR if we cannot merge this one.

**Comment by mshinwell** [comment]
@gasche This is in hand, and will be ready shortly.

File: asmcomp/cmmgen.ml
Comment (stedolan) [reply asmcomp/cmmgen.ml:1361]: I've removed the header.

I left in the low-bit check: I'm not convinced that it's correct to allow `Cload` to return an int with a zero in the low bit.

File: testsuite/tests/basic/switch_opts.ml
Comment (stedolan) [reply testsuite/tests/basic/switch_opts.ml]: It's not terribly important. I was imagining a bizarre failure mode where the table is generated but the constants are screwed up, and the test happens to pass because it doesn't actually examine the constants because they're the same object.

I've changed it to use `Sys.opaque_identity`

**Comment by mshinwell** [comment]
Sorry to pick nits, but two final things:
1. Please add the note about `Cload` as a comment, because it isn't obvious.
2. How about renaming the "const block" additions in this patch to "const table" or similar?  "Block" sounds like it should have a header.

OK with those changes.

**Comment by stedolan** [comment]
Nits picked

File: asmcomp/cmmgen.ml
Comment (gasche) [reply asmcomp/cmmgen.ml:1361]: > I left in the low-bit check: I'm not convinced that it's correct to allow Cload to return an int with a zero in the low bit.

Note that this is probably what happens in [strmatch.ml](https://github.com/ocaml/ocaml/blob/trunk/asmcomp/strmatch.ml#L76-L78), which comes from [MPR#6269](https://caml.inria.fr/mantis/view.php?id=6269). But I am completely fine with no trying it here -- I'm not sure why it would or wouldn't be correct either.

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
This has been cherry-picked to 4.05.

**Pull request merged**
